compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1840,7 +1840,11 @@ impl<'a> Parser<'a> {
                 if let token::Literal(token::Lit { kind: token::Integer, symbol, suffix }) =
                     next_token.kind
                 {
-                    if self.token.span.hi() == next_token.span.lo() {
+                    // If this integer looks like a float, then recover as such.
+                    if suffix.map_or(true, |s| s == sym::f32 || s == sym::f64)
+                        && symbol.as_str().chars().all(|c| c.is_numeric() || c == '_')
```

Comment:
Something that might deserve a comment: we will never encounter the exponent part of a floating point literal here, since the characters (`e` or `E`, and sometimes `+` or `-`) do not overlap with integer literals, so we only need to check that all characters are numeric.

Reply:
Everything from "we will" onwards would make a great comment here.